cmd/kubeadm/app/phases/upgrade/postupgrade.go

```diff
@@ -122,6 +123,30 @@ func WriteKubeletConfigFiles(cfg *kubeadmapi.InitConfiguration, patchesDir strin
 	}
 
 	errs := []error{}
+
+	// We will first check if instance-config.yaml exists,
+	// and if it doesn't, then we get the containerRuntimeEndpoint from kubeadm-flags.env and write to instance-config.yaml
+	if features.Enabled(cfg.FeatureGates, features.NodeLocalCRISocket) {
+		_, err := os.Stat(filepath.Join(kubeletDir, kubeadmconstants.KubeletInstanceConfigurationFileName))
+		if os.IsNotExist(err) {
+			containerRuntimeEndpoint, err := kubeletphase.ReadKubeadmFlagsFromFile(filepath.Join(kubeletDir, kubeadmconstants.KubeletEnvFileName), "container-runtime-endpoint")
```

Comment:
we also support `kubeadm upgrade ... --dry-run` on a non-existing cluster (i.e. this node has not joined a cluster).
this line will be problematic as the env file will not exist.

i think you should wrap the whole reading with `if !dryRun` and for `ifdryRun` maybe use a value like `containerRuntimeEndpoint:  unix://dry-run-endpoint`

Reply:
I changed the logic so that if containerRuntimeEndpoint is empty, we do not perform dry-run. Do you think this would be better?

Reply:
adding another comment about this in a follow up review.